File: backtracking/subsets.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> answer;
        vector<int> temp;

        backtrack(answer, temp, nums, 0);
        return answer;
    }

    void backtrack(vector<vector<int>>& answer, vector<int>& temp, vector<int>& nums, int start) {
        answer.push_back(temp);

        for(int i = start; i < nums.size(); i++) {
            temp.push_back(nums[i]);
            backtrack(answer, temp, nums, i + 1);
            temp.pop_back();
        }
    }
};

File: backtracking/subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        return subset(set(nums))
        
def subset(S):
    if len(S) == 0:
        return {frozenset({})}
    else:
        iter_S = iter(S)
        e = next(iter_S)
        
        T = S - {e}
        
        P_T = subset(T) 
        
        P_S = P_T | { frozenset(t | {e}) for t in P_T}
        
        return P_S